src/Symfony/Component/Config/Definition/EnumNode.php

```diff
@@ -78,13 +104,48 @@ protected function finalizeValue(mixed $value): mixed
     {
         $value = parent::finalizeValue($value);
 
-        if (!\in_array($value, $this->values, true)) {
-            $ex = new InvalidConfigurationException(\sprintf('The value %s is not allowed for path "%s". Permissible values: %s', json_encode($value), $this->getPath(), $this->getPermissibleValues(', ')));
-            $ex->setPath($this->getPath());
+        if ($this->enumFqcn) {
+            if (is_a($this->enumFqcn, \BackedEnum::class, true)) {
+                if (\is_string($value) || \is_int($value)) {
+                    try {
+                        $case = $this->enumFqcn::tryFrom($value);
+                    } catch (\TypeError) {
+                        throw new InvalidConfigurationException(\sprintf('The value could not be casted to a case of the "%s" enum. Is the value the same type as the backing type of the enum?', $this->enumFqcn));
+                    }
+
+                    if (null !== $case) {
+                        return $case;
+                    }
+                } elseif ($value instanceof \UnitEnum && !is_a($value, $this->enumFqcn)) {
+                    throw new InvalidConfigurationException(\sprintf('The value should be part of the "%s" enum, got a value from the "%s" enum.', $this->enumFqcn, get_debug_type($value)));
+                }
+            }
+
+            if (is_a($this->enumFqcn, \UnitEnum::class, true) && is_a($value, $this->enumFqcn, true)) {
```

Comment:
```suggestion
            if (null !== $this->enumFqcn && $value instanceof $this->enumFqcn) {
```

Reply:
Not sure why I didn't do this in first place. That's updated, thanks :+1: